packages/peregrine/src/Router/MagentoRouteHandler.js

```diff
@@ -1,74 +1,128 @@
+    componentDidUpdate() {
+        const { props, state } = this;
+        const { pathname } = props.location;
+        const isKnown = state.componentMap.has(pathname);
 
-        if (changed && !seen) {
-            this.getRouteComponent(nextProps.location.pathname);
+        if (!isKnown) {
+            this.getRouteComponent();
         }
     }
 
-    getRouteComponent(pathname) {
-        const { apiBase, __tmp_webpack_public_path__ } = this.props;
-
-        resolveUnknownRoute({
-            route: pathname,
-            apiBase,
-            __tmp_webpack_public_path__
-        })
-            .then(({ rootChunkID, rootModuleID, matched, id }) => {
-                if (!matched) {
-                    // TODO: User-defined 404 page
-                    // when the API work is done to support it
-                    throw new Error('404');
-                }
-                return fetchRootComponent(rootChunkID, rootModuleID).then(
-                    Component => {
-                        this.setState({
-                            [pathname]: {
-                                Component,
-                                id
-                            }
-                        });
-                    }
-                );
-            })
-            .catch(err => {
-                console.log('Routing resolve failed\n', err);
+    async getRouteComponent() {
+        const { __tmp_webpack_public_path__, apiBase, location } = this.props;
+        const { pathname } = location;
+
+        try {
+            // try to resolve the route
+            // if this throws, we essentially have a 500 Internal Error
+            const resolvedRoute = await resolveUnknownRoute({
+                __tmp_webpack_public_path__,
+                apiBase,
+                route: pathname
             });
+
+            const { id, matched, rootChunkID, rootModuleID } = resolvedRoute;
+
+            // if resolution and destructuring succeed but return no match
+            // then we have a straightforward 404 Not Found
+            if (!matched) {
+                throw new Error('404');
+            }
+
+            // at this point we should have a matching RootComponent
+            // if this throws, we essentially have a 500 Internal Error
+            const RootComponent = await fetchRootComponent(
+                rootChunkID,
+                rootModuleID
+            );
+
+            // associate the matching RootComponent with this location
+            this.setRouteComponent(pathname, RootComponent, { id });
+        } catch ({ message }) {
+            const symbol = message === '404' ? NotFound : InternalError;
+
+            // we don't have a matching RootComponent, but we've checked for one
+            // so associate the appropriate error case with this location
+            this.setRouteComponent(pathname, symbol);
+        }
+    }
+
+    setRouteComponent(pathname, RootComponent, meta) {
```

Comment:
Should it be `setRootcomponent`?

Reply:
Haha, I can see the confusion. It was named `setRouteComponent` because it's setting the component for the route. The prior function is `getRouteComponent,` which gets a component for the route.

Do you think they need to be renamed?

Reply:
We pronounce 'route' and 'root' the same here which has led to some confusion :smile: 
`setComponentToRoute` or something similar would clear up any confusion.